### PR TITLE
fix(enforcer): Apparmor policies not working as expected

### LINKED ISSUE #954: Apparmor allow based policies not working as expected.
## Description
```yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: autopol-system-1080768273
  namespace: default
spec:
  action: Allow
  file:
    matchDirectories:
    - dir: /
      fromSource:
      - path: /bin/bash
      recursive: true
    - dir: /etc/
      fromSource:
      - path: /bin/cat
      recursive: true
  process:
    matchPaths:
    - path: /bin/bash
    - path: /bin/cat
  selector:
    matchLabels:
      app: default
  severity: 1
```
Applying the following yaml policy should allow us to:
- run bash without seeing any error messages
- cat will get access only to /etc/**
## Acutual behavior
- Bash is working as expteced
- Executing cat command, yeilds a permission denied error.

# Proposed behavior for AppArmor allow based policies

We will be using apparmor domain transitions to enforce prolicies for each binary.

## White listed processes

For white listed processes we are going to enforce the following rules:
- If a process is allowed but no file access is granted, we will allow the process to run but no file/dir access is granted.

```yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: autopol-system-xxxxxxx
  namespace: default
spec:
  action: Allow
  process:
    matchPaths:
    - path: /bin/bash
    - path: /bin/cat
  selector:
    matchLabels:
      app: default
  severity: 1
```
The above policy should generate the following psedo-apparmor profile 

```
## == Managed by KubeArmor == ##

#include <tunables/global>

profile kubearmor-default-deploy01-c1 flags=(attach_disconnected,mediate_deleted) {
	#include <abstractions/base>

	network,
	capability,

	/bin/bash px -> kubearmor-default-deploy01-c1-XXXXX-bash,
    /bin/cat px -> kubearmor-default-deploy01-c1-XXXXX-cat,
}

profile kubearmor-default-deploy01-c1-XXXXX-bash flags=(attach_disconnected,mediate_deleted) {
}

profile kubearmor-default-deploy01-c1-XXXXX-cat flags=(attach_disconnected,mediate_deleted) {
}
```

- If from source is used to whitelist, a perticular process. The whitelisted process shoud NOT inherit access from the parent process.

```yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: autopol-system-xxxxxxx
  namespace: default
spec:
  action: Allow
  process:
    matchPaths:
    - path: /bin/cat
      fromSource:
      - path: /bin/bash
    - path: /bin/bash
  selector:
    matchLabels:
      app: default
  severity: 1
```
The above policy should generate the following psedo-apparmor profile 
```
## == Managed by KubeArmor == ##

#include <tunables/global>

profile kubearmor-default-deploy01-c1 flags=(attach_disconnected,mediate_deleted) {
	#include <abstractions/base>

	network,
	capability,

	/bin/bash px -> kubearmor-default-deploy01-c1-XXXXX-bash,
}

profile kubearmor-default-deploy01-c1-XXXXX-bash flags=(attach_disconnected,mediate_deleted) {
    /bin/cat px -> kubearmor-default-deploy01-c1-XXXXX-cat,
}

profile kubearmor-default-deploy01-c1-XXXXX-cat flags=(attach_disconnected,mediate_deleted) {
}
```

> bash should be explicitly defined in the whitelisted processes, else cat will never be able to run.

### PR DESCRIPTION
**Pull request by achrefbensaad**
**Purpose of PR?**:

Fixes #954 

**Does this PR introduce a breaking change?** No

**Checklist:**
- [x] Bug fix. Fixes #954 
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [x] Commit has unit tests
- [x] Commit has integration tests

### CODE REVIEW

**Comment by achrefbensaad** [comment]
TODO:

- [x] Add unit test to check if the same rule gets applied twice by two policies, nothing will break.
- [ ] Handle MatchPattern
Match pattern will be handled in a seperate PR.

**Comment by achrefbensaad** [comment]
image

Needs to be fixed.

Edit: Fixed

File: tests/smoke/res/ksp-two-policies-one-rule.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  namespace: wordpress-mysql
+spec:
+  severity: 5
+  message: "block /bin/cat from accessing /secret.txt"
+  selector:
+    matchLabels:
+      app: wordpress
+  file:
+    matchPaths:
+      - path: //etc/passwd
```
Comment (nyrahul) [inline tests/smoke/res/ksp-two-policies-one-rule.yaml]:
```suggestion
      - path: /etc/passwd
```

File: tests/smoke/res/ksp-two-policies-one-rule.yaml [hunk 2]
```diff
@@ -0,0 +1,37 @@
+apiVersion: security.kubearmor.com/v1
+kind: KubeArmorPolicy
+metadata:
+  name: ksp-ubutnu-1-file-path-block-from-source-1
```
Comment (nyrahul) [inline tests/smoke/res/ksp-two-policies-one-rule.yaml]:
```suggestion
  name: ksp-ubuntu-1-file-path-block-from-source-1
```

File: tests/smoke/res/ksp-two-policies-one-rule.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    Block
+---
+apiVersion: security.kubearmor.com/v1
+kind: KubeArmorPolicy
+metadata:
+  name: ksp-ubutnu-1-file-path-block-from-source-2
+  namespace: wordpress-mysql
+spec:
+  severity: 5
+  message: "block /bin/cat from accessing /secret.txt"
```
Comment (nyrahul) [inline tests/smoke/res/ksp-two-policies-one-rule.yaml]:
```suggestion
  message: "block /bin/cat from accessing /etc/passwd"
```

File: tests/smoke/res/ksp-two-policies-one-rule.yaml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      - path: //etc/passwd
+        fromSource:
+        - path: /bin/cat
+  action:
+    Block
+---
+apiVersion: security.kubearmor.com/v1
+kind: KubeArmorPolicy
+metadata:
+  name: ksp-ubutnu-1-file-path-block-from-source-2
```
Comment (nyrahul) [inline tests/smoke/res/ksp-two-policies-one-rule.yaml]:
```suggestion
  name: ksp-ubuntu-1-file-path-block-from-source-2
```

File: tests/smoke/res/ksp-two-policies-one-rule.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  severity: 5
+  message: "block /bin/cat from accessing /secret.txt"
+  selector:
+    matchLabels:
+      app: wordpress
+  file:
+    matchPaths:
+      - path: /etc/passwd
+        fromSource:
+        - path: /bin/cat
```
Comment (nyrahul) [inline tests/smoke/res/ksp-two-policies-one-rule.yaml]:
```suggestion
    matchPaths:
      - path: /etc/passwd
        fromSource:
        - path: /bin/cat
      - path: /etc/shadow
        fromSource:
        - path: /bin/cat
```

Comment (nyrahul) [reply tests/smoke/res/ksp-two-policies-one-rule.yaml]: The aim is to not have exactly the same rule but a combination of rules.

**Review by nyrahul — DISMISSED** [review]
Brilliant PR. Mostly LGTM. Request few changes in test cases.

**Comment by daemon1024** [comment]
Glad to see that all the changes are contained in the template 😬 

@achrefbensaad do you have an image which I can try this out with?

**Comment by achrefbensaad** [comment]
@daemon1024 , you can use ttl.sh/kubearmor-apparmor

**Comment by achrefbensaad** [comment]
@nyrahul @daemon1024 , I have edited the PR and introduced unloading of unsued apparmor profiles.

**Review by Ankurk99 — APPROVED** [review]
Awesome work @achrefbensaad. I tried the changes and it's running smooth

**Review by daemon1024 — APPROVED** [review]

**Pull request merged**